main/touched.zkasm

```diff
@@ -0,0 +1,120 @@
+VAR GLOBAL tmpB
+VAR GLOBAL tmpC
+VAR GLOBAL tmpD
+VAR GLOBAL tmpE
+VAR GLOBAL tmpStateRoot
+
+initTouchedTree:
+    0               :MSTORE(touchedSR)
+                    :RETURN
+
+; @info Save touched root when a new context is created
+commitTouched:
+    $ => A          :MLOAD(touchedSR)
+    A               :MSTORE(initTouchedSR)
+                    :RETURN
+
+; @info Back to initial touched root
+revertTouched:
+    $ => A               :MLOAD(initTouchedSR)
+    A                    :MSTORE(touchedSR)
+
+; @info Check if an address is warmed. If it is not, add it to the touched tree
+; @in A => address
+; @out D => if the address is warmed [0 if warm, 1 if cold]
+isWarmedAddress:
+    ; store previous registers values temporary
+    B               :MSTORE(tmpB)
+    C               :MSTORE(tmpC)
+    E               :MSTORE(tmpE)
+
+    ; if address is a zero address check if it is warm
+    0 => B
+    $                           :EQ, JMPC(checkWarmed)
+    ; if address is a precompiled considered as warm address
+    10 => B
+    $                           :LT, JMPC(finishWarmPrecompiled)
+
+checkWarmed:
+    ; save current state root load touched root
+    SR              :MSTORE(tmpStateRoot)
+    $ => SR         :MLOAD(touchedSR)
```

Comment:
SR for touched slots and touched address should be different, the values can collide!

Reply:
they cannot collide since they have different constant to compute the key of the tree